packages/bundle/esbuild.static.mjs

```diff
@@ -0,0 +1,266 @@
+
+  return work(cwd);
+}
+
+async function addConfig(
+  /** @type { import('esbuild').OnResolveArgs } */
+  { path, resolveDir }
+) {
+  const importPathURL = importMetaResolve(path, pathToFileURL(resolveDir) + '/');
+
+  if (!importPathURL.startsWith('file:')) {
+    return;
+  }
+
+  const {
+    packageJson: { name, version },
+    path: packagePath
+  } = await readPublishingPackageJSONUp(fileURLToPath(importPathURL));
+
+  const fullName = `${name}@${version}`;
+
+  if (!path.startsWith(name)) {
+    throw new Error(`args.path must starts with name, args.path = ${path}, name = ${name}`);
+  }
+
+  const familyName = `${flatName(name)}__${version}`;
+  const entryNames = `${familyName}__[name]`;
+
+  let currentConfig = configs.get(fullName);
+
+  if (!currentConfig) {
+    /** @type { import('esbuild').BuildOptions } */
+    currentConfig = {
+      absWorkingDir: dirname(packagePath),
+      chunkNames: `${familyName}__[name]-[hash]`,
+      entryNames,
+      entryPoints: {}
+    };
+
+    configs.set(fullName, currentConfig);
+  }
+
+  const entries = new Map(Object.entries(currentConfig.entryPoints));
+
+  const [moduleName, namedExports] = extractPackageNameAndNamedExports(path);
+  const entryName = flatName(namedExports || moduleName.split('/').at(-1));
+
+  if (!entries.has(entryName)) {
+    entries.set(entryName, path);
+
+    currentConfig.entryPoints = Object.fromEntries(Array.from(entries.entries()));
+    currentConfig.write = true;
+  }
+
+  return `./${entryNames.replace(/\[name\]/gu, entryName)}.js`;
+}
+
+function getPendingConfig() {
+  for (const value of configs.values()) {
+    if (value.write !== false) {
+      return value;
+    }
+  }
+}
+
+const BASE_CONFIG = {
+  alias: {
+    adaptivecards: '@msinternal/adaptivecards',
+    'base64-js': '@msinternal/base64-js',
+    'botframework-directlinejs': '@msinternal/botframework-directlinejs',
+    'html-react-parser': '@msinternal/html-react-parser',
+    'microsoft-cognitiveservices-speech-sdk': '@msinternal/microsoft-cognitiveservices-speech-sdk',
+    'object-assign': '@msinternal/object-assign',
+    'react-is': '@msinternal/react-is'
+  },
+  bundle: true,
+  external: ['react', 'react-dom'],
+  format: 'esm',
+  loader: { '.js': 'jsx' },
+  minify: true,
+  outdir: resolve(fileURLToPath(import.meta.url), `../static/`),
+  platform: 'browser',
+  sourcemap: true,
+  splitting: true,
+  write: true,
+
+  /** @type { import('esbuild').Plugin[] } */
+  plugins: [
+    {
+      name: 'static-builder',
+      setup(build) {
+        // eslint-disable-next-line require-unicode-regexp
+        build.onResolve({ filter: /^[^.]/ }, async args => {
+          // "external" field only works if the plug-in give up (return undefined.)
+          if (args.path === 'react' || args.path === 'react-dom') {
+            return undefined;
+          }
+
+          // Only ESM can be externalized, CJS cannot be externalized because require() is not guaranteed to be at top-level.
```

Comment:
I assume this is esbuild limitation. We can try work around it by using different transform from CJS to ESM (there are a lot of options available now).

Maybe we can bail if `require` is found not in top-level.

Reply:
Think about `a && require('b')` where `b` has side effects. We can't really convert this into `import`.

But it's relatively rare.